### ISSUE TITLE
BUG: sort() with order slow on multi dtype arrays (sorting by unsigned integer ID)

### ISSUE THREAD

**Issue by me9hanics** [numpy/numpy#27007]
### Describe the issue:

I'm using arrays with specified datatypes. I noticed that sorting such an array with `np.sort(..., order='id')`seems to be very slow. This comes as a surprise, as if we just sort the ID column separately with `np.sort` it is handled fastly, and no other column is relevant for the sorting, therefore using `order` here should be comparably fast to just sorting a 1D array.

If I run `np.sort` with order by column as in the example, it takes around ~20 seconds on my machine. Compared to just sorting the column, which runs in 1/100th of the time, this seems unreasonably slow. One could argue that it is more complicated to sort multiple column and datatype arrays, but this way it'd be faster to just take the order column, sort it, get the permutation and order all columns with the same permutation.

I used numpy 2.0.1 to test it (Python 3.12.4), but it is an issue with older versions too, e.g. 3.12.4.

### Reproduce the code example:

```python
import numpy as np
import time
dtypes = [('x', 'u1'), ('id', '<u8')]
x_values = [n % 2 for n in range(5000000)]
ids = list(range(5000000, 0, -1))
x = np.array(list(zip(x_values, ids)), dtype=dtypes)

t1 = time.time()
multi_dtype_array_sort = np.sort(x, order='id')
t2 = time.time()
time_1 = t2 - t1

t1 = time.time()
sort_1D = np.sort(x['id'])
t2 = time.time()
time_2 = t2 - t1

print(time_1, time_2)
```

### Python and NumPy Versions:

Python: 3.12.4
Numpy: 2.0.1

### Context for the issue:

(I'm partially working with embedded code, hence the specified datatypes, these come from outputs of a sensor. I need to sort the array by the ID column.)

**Comment by me9hanics** [comment]
> I'm not clear on what you are doing, but if you want to sort the entire array using only one column, using argsort on the single column followed by a take might be faster, as the sort only needs to move around integers rather than whole rows.

Yes, that is what I do instead, kind of hinted on it with the last sentence in the 2nd paragraph - just wanted to point out that the `order` option seems to be implemented in not an efficient way

**Comment by Siddharth-Latthe-07** [comment]
The slower performance might be due to additional overhead involved in managing the structured array's metadata and handling the sorting process internally, compared to directly sorting a 1D array. The sort operation on the structured array might be performing unnecessary operations or not taking full advantage of potential optimizations that could be applied when sorting by a single field.
As a workaround, you can sort the 'id' field separately, obtain the sorted indices, and then apply these indices to reorder the entire structured array. This approach should be more efficient:
```
import numpy as np
import time

# Define the data types for the structured array
dtypes = [('x', 'u1'), ('id', '<u8')]

# Create data for the array
x_values = [n % 2 for n in range(5000000)]
ids = list(range(5000000, 0, -1))

# Create the structured array
x = np.array(list(zip(x_values, ids)), dtype=dtypes)

# Measure the time to sort the 'id' field separately and reorder the array
t1 = time.time()
sorted_indices = np.argsort(x['id'])
sorted_x = x[sorted_indices]
t2 = time.time()
time_3 = t2 - t1

print(f"Time to sort structured array using workaround: {time_3:.5f} seconds")
```
Hope, this helps
Thanks

**Comment by seberg** [comment]
@Siddharth-Latthe-07 could you please slow down with these posts, it reminds me of LLM use whether true or not.

It is nice that you are trying to help, but e.g. here your awnser did not add any information the workaround was already mentioned twice and that the slow-down is related to structureed arrays is clear.

**Comment by charris** [comment]
Structured arrays are always sorted on all fields, the order argument just changes the order of the sorts. That probably accounts for the slownest.

**Comment by seberg** [comment]
Yeah, there are two things.  One is that you should just expect them to be slower, since one deals with a single structured entry and that must add logic.
The other is that sorting (unlike casts, take, etc.) doesn't (and likely cannot without reformulating sorting) prepare the comparison/data movement, so it will have a huge overhead each time to figure out what the structure is (which isn't encoded in a quick way).
